// @tbdex/protocol/dist/esm/src/message-kinds/order-status.js
import { Message } from '../message.js';
/**
 * Sent by the PFI to Alice to convey the current status of an order. There can be many OrderStatus
 * messages in a given Exchange
 * @beta
 */
export class OrderStatus extends Message {
    /** a set of valid Message kinds that can come after an order status */
    validNext = new Set([]);
    /**
     * Creates an order status with the given options
     * @param opts - options to create an order status
     */
    static create(opts) {
        const metadata = {
            ...opts.metadata,
            kind: 'orderstatus',
            id: Message.generateId('orderstatus'),
            createdAt: new Date().toISOString()
        };
        const message = { metadata, data: opts.data };
        Message.validateData('orderstatus', message.data);
        return new OrderStatus(message);
    }
    /** Current status of Order that's being executed (e.g. PROCESSING, COMPLETED, FAILED etc.) */
    get orderStatus() {
        return this.data.orderStatus;
    }
}
