// @tbdex/protocol/dist/cjs/src/resource.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { typeid } from 'typeid-js';
import { Crypto } from './crypto.js';
import { validate } from './validator.js';
/**
 * tbDEX Resources are published by PFIs for anyone to consume and generally used as a part of the discovery process.
 * They are not part of the message exchange, i.e Alice cannot reply to a Resource.
 * @beta
 */
var Resource = /** @class */ (function () {
    /**
     * Constructor is primarily for intended for internal use. For a better developer experience,
     * consumers should use concrete classes to programmatically create resources (e.g. Offering class) and
     * {@link Resource.parse} to parse stringified resources.
     * @param jsonResource - the resource as a json object
     * @param data - `resource.data` as a ResourceKind class instance. can be passed in as an optimization if class instance
     * is present in calling scope
     */
    function Resource(jsonResource) {
        this._metadata = jsonResource.metadata;
        this._data = jsonResource.data;
        this._signature = jsonResource.signature;
    }
    /**
     * parses the json resource into a Resource instance. performs format validation and an integrity check on the signature
     * @param payload - the resource to parse. can either be an object or a string
     */
    Resource.parse = function (resource) {
        return __awaiter(this, void 0, void 0, function () {
            var jsonResource, errorMessage;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        try {
                            jsonResource = typeof resource === 'string' ? JSON.parse(resource) : resource;
                        }
                        catch (e) {
                            errorMessage = e instanceof Error ? e.message : e;
                            throw new Error("parse: Failed to parse resource. Error: ".concat(errorMessage));
                        }
                        return [4 /*yield*/, Resource.verify(jsonResource)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, Resource.factory(jsonResource)];
                }
            });
        });
    };
    /**
     * validates the resource and verifies the cryptographic signature
     * @throws if the message is invalid
     * @throws see {@link Crypto.verify}
     * @returns Resource signer's DID
     */
    Resource.verify = function (resource) {
        return __awaiter(this, void 0, void 0, function () {
            var jsonResource, digest, signerDid;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        jsonResource = resource instanceof Resource ? resource.toJSON() : resource;
                        Resource.validate(jsonResource);
                        digest = Crypto.digest({ metadata: jsonResource.metadata, data: jsonResource.data });
                        return [4 /*yield*/, Crypto.verify({ detachedPayload: digest, signature: jsonResource.signature })];
                    case 1:
                        signerDid = _a.sent();
                        if (jsonResource.metadata.from !== signerDid) { // ensure that DID used to sign matches `from` property in metadata
                            throw new Error('Signature verification failed: Expected DID in kid of JWS header must match metadata.from');
                        }
                        return [2 /*return*/, signerDid];
                }
            });
        });
    };
    /**
     * validates the resource provided against the appropriate json schemas.
     * 2-phased validation: First validates the resource structure and then
     * validates `data` based on the value of `metadata.kind`
     * @param jsonResource - the resource to validate
     *
     * @throws `Error` if validation fails
     */
    Resource.validate = function (jsonResource) {
        validate(jsonResource, 'resource');
        Resource.validateData(jsonResource['metadata']['kind'], jsonResource['data']);
    };
    /**
     * Validates `data` section of resource only. This is useful for partially validating
     * unsigned Resources.
     */
    Resource.validateData = function (kind, resourceData) {
        // validate the value of `data`
        validate(resourceData, kind);
    };
    /** Generates a unique id with the resource kind's prefix */
    Resource.generateId = function (resourceKind) {
        return typeid(resourceKind).toString();
    };
    /**
     * signs the message as a jws with detached content and sets the signature property
     * @param did - the signer's DID
     */
    Resource.prototype.sign = function (did) {
        return __awaiter(this, void 0, void 0, function () {
            var payload, payloadDigest, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        payload = { metadata: this.metadata, data: this.data };
                        payloadDigest = Crypto.digest(payload);
                        _a = this;
                        return [4 /*yield*/, Crypto.sign({ did: did, payload: payloadDigest, detached: true })];
                    case 1:
                        _a._signature = _b.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * validates the resource and verifies the cryptographic signature
     * @throws if the resource is invalid
     * @throws see {@link Crypto.verify}
     * @returns Resource signer's DID
     */
    Resource.prototype.verify = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, Resource.verify(this)];
            });
        });
    };
    /**
     * returns the message as a json object. Automatically used by `JSON.stringify` method.
     */
    Resource.prototype.toJSON = function () {
        return {
            metadata: this.metadata,
            data: this.data,
            signature: this.signature
        };
    };
    Object.defineProperty(Resource.prototype, "metadata", {
        /** The metadata object contains fields about the resource and is present in every tbdex resource. */
        get: function () {
            return this._metadata;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Resource.prototype, "data", {
        /** the actual resource kind's content data */
        get: function () {
            return this._data;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Resource.prototype, "signature", {
        /** the resource's cryptographic signature */
        get: function () {
            return this._signature;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Resource.prototype, "id", {
        /** the resource's id */
        get: function () {
            return this.metadata.id;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Resource.prototype, "kind", {
        /** the resource kind (e.g. offering) */
        get: function () {
            return this.metadata.kind;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Resource.prototype, "from", {
        /** The sender's DID */
        get: function () {
            return this.metadata.from;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Resource.prototype, "createdAt", {
        /** Resource creation time. Expressed as ISO8601 */
        get: function () {
            return this.metadata.createdAt;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Resource.prototype, "updatedAt", {
        /** Resource last updated time. Expressed as ISO8601 */
        get: function () {
            return this.metadata.updatedAt;
        },
        enumerable: false,
        configurable: true
    });
    /** offering type guard */
    Resource.prototype.isOffering = function () {
        return this.metadata.kind === 'offering';
    };
    return Resource;
}());
export { Resource };
